toolkits/salesforce/arcade_salesforce/models.py

```diff
@@ -0,0 +1,344 @@
+            response = await client.get(**kwargs)
+
+            if response.status_code >= 300:
+                self._raise_http_error(response)
+
+            return cast(dict, response.json())
+
+    async def post(
+        self,
+        endpoint: str,
+        data: Optional[dict] = None,
+        json_data: Optional[dict] = None,
+        headers: Optional[dict] = None,
+    ) -> dict:
+        async with httpx.AsyncClient() as client:
+            kwargs: dict[str, Any] = {
+                "url": self._endpoint_url(endpoint),
+                "headers": self._build_headers(headers),
+            }
+            if data:
+                kwargs["data"] = data
+            if json_data:
+                kwargs["json"] = json_data
+
+            response = await client.post(**kwargs)
+
+            if response.status_code >= 300:
+                self._raise_http_error(response)
+
+        return cast(dict, response.json())
+
+    async def get_object_fields(self, object_type: SalesforceObject) -> list[str]:
+        if object_type not in self.object_fields:
+            response = await self._describe_object(object_type)
+            self.object_fields[object_type] = [field["name"] for field in response["fields"]]
+
+        return self.object_fields[object_type]
+
+    async def _describe_object(self, object_type: SalesforceObject) -> dict:
+        return await self.get(f"sobjects/{object_type.value}/describe/")
+
+    async def _get_related_objects(
+        self,
+        child_object_type: SalesforceObject,
+        parent_object_type: SalesforceObject,
+        parent_object_id: str,
+        limit: Optional[int] = 10,
+    ) -> list[dict]:
+        try:
+            response = await self.get(
+                f"sobjects/{parent_object_type.value}/{parent_object_id}/{child_object_type.plural.lower()}",
+                params={"limit": limit},
+            )
+            return cast(list, response["records"])
+        except ResourceNotFoundError:
+            return []
+
+    async def get_object_by_id(self, object_id: str) -> Optional[dict]:
+        try:
+            response = await self.get(f"sobjects/{object_id}")
+            return clean_object_data(response)
+        except ResourceNotFoundError:
+            if "User" not in object_id:
+                return await self.get_object_by_id(f"User/{object_id}")
+            return None
+
+    async def get_account(self, account_id: str) -> Optional[dict[str, Any]]:
+        try:
+            return cast(dict, await self.get(f"sobjects/Account/{account_id}"))
+        except ResourceNotFoundError:
+            return None
+
+    async def get_account_contacts(self, account_id: str, limit: Optional[int] = 10) -> list[dict]:
+        contacts = await self._get_related_objects(
+            SalesforceObject.CONTACT, SalesforceObject.ACCOUNT, account_id, limit
+        )
+
+        return [
+            clean_contact_data(contact)
+            for contact in await asyncio.gather(*[
```

Comment:
I worry about getting rate limited with this sort of stuff. Although it decreases tool execution time, it increases the risk of being rate limited.

If we are invoking multiple API calls concurrently like we are here, then implementing a rate limiting strategy would be preferred

Reply:
I added a [semaphore in the `SalesforceClient` class](https://github.com/ArcadeAI/arcade-ai/blob/34742b4970dc189f87665dd83a8881f3c757e126/toolkits/salesforce/arcade_salesforce/models.py#L40) limiting concurrency. The [default is 3 requests, configurable through an env var](https://github.com/ArcadeAI/arcade-ai/blob/34742b4970dc189f87665dd83a8881f3c757e126/toolkits/salesforce/arcade_salesforce/constants.py#L5).